File: 백준/Gold/1753. 최단경로/최단경로.py
import heapq
import sys
input = sys.stdin.readline

V,E = map(int,input().split())
K = int(input())
INF = 1e9
#무한으로 초기화
dp = [INF] * (V+1)
h = []
graph = [[] for _ in range(V+1)]

def Dijkstra(s):
    #시작 정점은 0으로 초기화
    dp[s] = 0
    #가중치와 시작점
    heapq.heappush(h,(0,s))

    while h:
        weight,n = heapq.heappop(h)
        #가중치가 더 크면 무시
        if dp[n] < weight:
            continue

        for w,next_node in graph[n]:
            next_weight = w+weight
            if next_weight < dp[next_node]:
                dp[next_node] = next_weight
                heapq.heappush(h,(next_weight,next_node))

for i in range(E):
    u,v,w = map(int,input().split())
    graph[u].append((w,v))

Dijkstra(K)
for i in range(1,V+1):
    print("INF" if dp[i] == INF else dp[i])

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static class Edge {
        int to, weight;

        public Edge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }

    static List<List<Edge>> graph;
    static int[] distance;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int V = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());

        int K = Integer.parseInt(br.readLine());

        graph = new ArrayList<>();
        for (int i = 0; i <= V; i++) {
            graph.add(new ArrayList<>());
        }

        distance = new int[V + 1];
        Arrays.fill(distance, Integer.MAX_VALUE);

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            graph.get(u).add(new Edge(v, w));
        }

        dijkstra(V, K);

        for (int i = 1; i <= V; i++) {
            if (distance[i] == Integer.MAX_VALUE)
                System.out.println("INF");
            else
                System.out.println(distance[i]);
        }
    }

    private static void dijkstra(int V, int start) {
        PriorityQueue<Edge> pq = new PriorityQueue<>(Comparator.comparingInt(e -> e.weight));
        distance[start] = 0;
        pq.offer(new Edge(start, 0));

        while (!pq.isEmpty()) {
            Edge current = pq.poll();
            int from = current.to;
            int dist = current.weight;

            if (dist > distance[from]) continue;

            for (Edge next : graph.get(from)) {
                int to = next.to;
                int nextDist = dist + next.weight;

                if (nextDist < distance[to]) {
                    distance[to] = nextDist;
                    pq.offer(new Edge(to, nextDist));
                }
            }
        }
    }
}